lib/fast_jsonapi/serialization_core.rb

```diff
@@ -65,46 +66,90 @@ def meta_hash(record, params = {})
+        "#{namespace}-#{record.cache_key}"
+      end
+
       def record_hash(record, fieldset, includes_list, params = {})
         if cache_store_instance
-          cache_opts = record_cache_options(cache_store_options, fieldset, includes_list, params)
-          record_hash = cache_store_instance.fetch(record, **cache_opts) do
-            temp_hash = id_hash(id_from_record(record, params), record_type, true)
-            temp_hash[:attributes] = attributes_hash(record, fieldset, params) if attributes_to_serialize.present?
-            temp_hash[:relationships] = relationships_hash(record, cachable_relationships_to_serialize, fieldset, includes_list, params) if cachable_relationships_to_serialize.present?
-            temp_hash[:links] = links_hash(record, params) if data_links.present?
-            temp_hash
+          cache_opts = record_cache_options(record, cache_store_options, fieldset, includes_list, params)
+          name_space = cache_opts.delete(:namespace)
+          cache_key = generate_cache_key(record, name_space)
+          klass = self
+          param = {
+            cache_key: cache_key,
+            cache_opts: cache_opts,
+            record: record,
+            klass: klass,
+            params: params
+          }
+          ::BatchLoader.for(param).batch do |batch_params, loader|
+            cache_keys = batch_params.map { |h| h[:cache_key] }
+            # load the cached value from cache store
+            cache_hits = cache_store_instance.read_multi(cache_keys)
+
+            # for not cached record, group it by cache options
+            uncached_by_opts = batch_params
+              .reject { |h| cache_hits.keys.include? h[:cache_key] }
+              .group_by { |h| h[:cache_opts] }
+            # replace grouped batch param with Hash: { cache_key_1 => record_hash_1, ... }
+            uncached_by_opts.transform_values! do |arr|
+              arr.each_with_object({}) do |b_param, record_hash_pairs|
+                temp_hash = b_param[:klass].id_hash(b_param[:klass].id_from_record(b_param[:record], b_param[:params]), b_param[:klass].record_type, true)
+                temp_hash[:attributes] = b_param[:klass].attributes_hash(b_param[:record], fieldset, b_param[:params]) if b_param[:klass].attributes_to_serialize.present?
+                temp_hash[:relationships] = b_param[:klass].relationships_hash(b_param[:record], nil, fieldset, includes_list, b_param[:params]) if b_param[:klass].relationships_to_serialize.present?
+                temp_hash[:links] = b_param[:klass].links_hash(b_param[:record], b_param[:params]) if b_param[:klass].data_links.present?
+                temp_hash[:meta] = b_param[:klass].meta_hash(b_param[:record], b_param[:params]) if b_param[:klass].meta_to_serialize.present?
+                record_hash_pairs[b_param[:cache_key]] = temp_hash
+              end
+            end
+            # cache the uncached record
+            uncached_by_opts.each do |opts, record_hash_pairs|
```

Comment:
`record_hash_pairs` is misleading - this is not a pair of record hashes, but rather `record_hashes_by_cache_key `

Reply:
```suggestion
            uncached_by_opts.each do |cache_opts, record_hashes_by_cache_key |
```